exporters/failure/collector_jira.py

```diff
@@ -41,108 +38,128 @@
+        apikey : str
+            API Token of your user. To create one, visit
+            https://id.atlassian.com/manage-profile/security/api-tokens
+        server : str
+            The server your JIRA is hosted.
+        projects : Optional[str]
+            The JIRA projects, separated by commas (,), if any.
+        """
         super().__init__(server, user, apikey)
+        self.projects = projects
+        self.jira_resolved_statuses = pelorus.utils.get_env_var(RESOLVED_STATUS_ENV)
         self.jql_query_string = pelorus.utils.get_env_var(
             JQL_SEARCH_QUERY_ENV, DEFAULT_JQL_SEARCH_QUERY
         )
-        self.jira_resolved_statuses = pelorus.utils.get_env_var(RESOLVED_STATUS_ENV)
-
-        self.query_result_fields_string = QUERY_RESULT_FIELDS
-
-        # Do not mix projects with custom JQL query
-        # Gather all fields and projects
-        if self.jql_query_string != DEFAULT_JQL_SEARCH_QUERY:
-            self.query_result_fields_string = ""
-        else:
-            if projects is not None and len(projects) > 0:
-                projects_str = '","'.join(projects.split(","))
-                self.jql_query_string = (
-                    self.jql_query_string
-                    + ' AND project in ("{}")'.format(projects_str)
-                )
+        self.query_result_fields_string = (
+            ""
+            if self.jql_query_string != DEFAULT_JQL_SEARCH_QUERY
+            else QUERY_RESULT_FIELDS
+        )
 
     def _connect_to_jira(self) -> JIRA:
-        """Method to connect to JIRA instance which may be cloud based
-        or self-hosted.
-        """
-        jira_client = None
-        server = self.server
-        email = self.user
-        api_key = self.apikey
-
-        # Connect to Jira
-        jira_client = JIRA(options={"server": server}, basic_auth=(email, api_key))
-
-        # Ensure connection was performed
+        """Connect to JIRA instance which may be cloud based or self-hosted."""
         try:
+            # Connect to JIRA
+            jira_client = JIRA(
+                options={"server": self.server}, basic_auth=(self.user, self.apikey)
+            )
+            # Ensure connection was performed
             jira_client.session()
+            return jira_client
         except JIRAError as error:
             logging.error(
                 "Status: %s, Error Response: %s", error.status_code, error.text
             )
             raise
 
-        return jira_client
+    def _set_jql_queries_strings(self, jira_client: JIRA) -> None:
+        """
+        Set JQL query string for projects.
+
+        Parameters
+        ----------
+        jira_client : JIRA
+            JIRA instance.
+        """
+        if self.jql_query_string == DEFAULT_JQL_SEARCH_QUERY and self.projects:
+            _projects = set(self.projects.split(","))
+            existing_projects = {project.name for project in jira_client.projects()}
```

Comment:
I do wonder about performance impact here. If API allows to access many projects it may be heavy on that call?

Reply:
Good point, I did not test and do not know the answer. I think we can test it out in a big example. If it is heavy, I believe we could change this part (lines 102 until 109) to
```python
            projects_str = ""
            for project in _projects:
                try:
                    jira_client.project(project)
                    projects_str += f'","{project}' if projects_str else project
                except JIRAError:
                    logging.error(
                        'Project of name "%s" does not exist in server %s',
                        project,
                        self.server,
                    )
```
I did not test if the `project` function works passing the name of the JIRA project.

Reply:
@mateusoliveira43 or even something else. Keep the previous logic (get all of the specified projects) and only if it fails apply the above. Above loop is much nicer then getting all imo.

Reply:
Alright, I will change it

Reply:
Here there is a problem: to use the `project` function we need to pass the key of the project, not its name. For example: in your example we have to pass "FIRST" instead of "Test". What you suggest to overcome this?

Reply:
I used the error message to get them, I do not like this solution because if the error message is updated by the the JIRA Python library, our code breaks.

Reply:
I think it's what we need to live with. Possibly some test to cover such scenario?

Here is a bit of code that is ugly and of course could be in recurrent function to have only logic once,  but falls-back to per project query if something goes wrong (maybe err 400 is enough here?):

```
from jira import JIRA
from jira.exceptions import JIRAError

api_key="***************"
email="***************"
server="https://pelorustest.atlassian.net"

projects_str="TEST,MyProject"

jql_query_string = ('type in ("Bug") AND priority in ("Highest") AND project in ("{}")').format(projects_str)

jira_client = JIRA(options={"server": server}, basic_auth=(email, api_key))
jira_search_results = []

try:
    jira_search_results = jira_client.search_issues(jql_query_string, startAt=0, maxResults=50, fields="summary",)
except JIRAError as jira_error:
    if jira_error.status_code == 400 and \
        "does not exist for the field 'project'" in jira_error.text:
        for project in projects_str.split(','):
            jql_query_string = ('type in ("Bug") AND priority in ("Highest") AND project in ("{}")').format(project.strip())
            try:
                jira_search_results += jira_client.search_issues(jql_query_string, startAt=0, maxResults=50, fields="summary",)
            except JIRAError as jira_error:
                if jira_error.status_code == 400 and \
                    "does not exist for the field 'project'" in jira_error.text:
                    pass

jira_search_results

```

Reply:
I implement something just like that, just added a logic to do the query maximum 2 times, instead of 1 + `n` (`n` being the number of projects passed by the user).

Yes, I still need to write tests, and I forgot to apply the `strip` function to the strings